src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.targets

```diff
@@ -92,12 +92,21 @@ Copyright (c) .NET Foundation. All rights reserved.
                       RuntimeIdentifier="$(RuntimeIdentifier)"
                       PlatformLibraryName="$(MicrosoftNETPlatformLibrary)"
                       CompilerOptions="@(DependencyFileCompilerOptions)">
-
-      <Output TaskParameter="FilesWritten" ItemName="FileWrites" />
     </GenerateDepsFile>
 
   </Target>
 
+  <!-- Use a separate target to add the deps file to the FileWrites item group, so that it is still included even if the 
+       GenerateDepsFile target is skipped due to being up-to-date in an incremental build. -->
+  <Target Name="CheckForBuildDependencyFileOutput"
+          DependsOnTargets="GenerateBuildDependencyFile"
+          BeforeTargets="_CheckForCompileOutputs"
+          Condition=" '$(GenerateDependencyFile)' == 'true'">
+    <ItemGroup>
+      <FileWrites Include="$(ProjectDepsFilePath)" Condition="Exists('$(ProjectDepsFilePath)')"/>
```

Comment:
Don't put this in a separate target, add it to `GenerateBuildDependencyFile`.

MSBuild will manipulate items that it can statically understand even when it skips a target for being up to date. It wasn't doing that before because it can't predict what the task would have output.